client/src/components/geneExpression/index.js

```diff
@@ -138,27 +160,43 @@ class GeneExpression extends React.Component {
       Apod,,, Cd74,,    ,,,    Foo,    Bar-2,,
     */
     if (bulkAdd !== "") {
+      const column = world.varAnnotations.col(varIndexName);
+
       const genes = _.pull(_.uniq(bulkAdd.split(/[ ,]+/)), "");
+      const worldGenes = column.asArray();
+
+      // These gene lists are unique enough where memoization is useless
+      const upperGenes = this._genesToUpper(genes);
+      const upperUserDefinedGenes = this._genesToUpper(userDefinedGenes);
+
+      const upperWorldGenes = this._memoGenesToUpper(column);
 
       dispatch({ type: "bulk user defined gene start" });
-      Promise.all(
-        genes.map(gene => {
-          if (gene.length === 0) {
-            return keepAroundErrorToast("Must enter a gene name.");
-          }
-          if (userDefinedGenes.indexOf(gene) !== -1) {
-            return keepAroundErrorToast("That gene already exists");
-          }
-          if (
-            world.varAnnotations.col(varIndexName).indexOf(gene) === undefined
-          ) {
-            return keepAroundErrorToast(
-              `${gene} doesn't appear to be a valid gene name.`
-            );
-          }
-          return dispatch(actions.requestUserDefinedGene(gene));
-        })
-      ).then(
+
+      const promises = [];
+      for (let i = 0, { length } = upperGenes; i < length; i += 1) {
+        const upperGene = upperGenes[i];
+
+        if (genes.length === 0) {
```

Comment:
this should not be in the loop - it is a constant every time.   But by putting it in the loop, this necessary check is omitted.   Move outside the loop, eg, immeiately after the `genes` variable is set (line 165).